src/main/java/net/minecraftforge/common/FarmlandWaterManager.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2018.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common;
+
+import com.google.common.base.Preconditions;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.World;
+import net.minecraftforge.common.ticket.AABBTicket;
+import net.minecraftforge.common.ticket.SimpleTicket;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+public class FarmlandWaterManager
+{
+    private static final Int2ObjectMap<Set<SimpleTicket<Vec3d>>> customWaterHandler = new Int2ObjectOpenHashMap<>();
```

Comment:
This is just asking for a massive seek time when large worlds are loaded.
Perhaps break this down into chunk based maps so that it can have a more consistant lookup time no matter the wise the world grows to.
You'd need to deal with the chunk load/unload events. As well as adding a static 'maxRange' to the tickets. Allowing a ticket to change sizes is a bad idea in my option. Static size to get down to the chunk level, and then matches to check the specific block.